core/client/index.ts

```diff
@@ -101,12 +149,97 @@ export class Core {
+      // field (holding the actual string message), as well as an optional
+      // validation payload encoded in the "validationSeq" and
+      // "validationDigest" fields. While we are implementing some explicit
+      // sanity checks, wrap the entire decoding implementation in a try block
+      // to handle any (unexpected) failures gracefully.
+      let payload = JSON.parse(decPayload);
+
+      if (!("message" in payload)
+          || (typeof payload["message"] !== "string"
+              && !(payload["message"] instanceof String))) {
+        throw "Decrypted message does not have a string payload.";
+      }
+      messagePayload = String(payload["message"]).valueOf();
+
+      recipients = payload["recipients"];
+      if (!Array.isArray(recipients)) {
+        throw "Message recipients is not an array.";
+      }
+      for (let r of recipients) {
+        if (typeof r !== "string" && !(r instanceof String)) {
+          throw "Message recipient is not a string.";
+        }
+      }
+
+      // If one of the validation payload fields is present, try to decode it as
+      // a validation payload. This explicitly provokes an error when just one
+      // of the required fields is provided.
+      if ("validationSeq" in payload || "validationDigest" in payload) {
+        let validationSeq = payload["validationSeq"];
+        let validationDigest = payload["validationDigest"];
+
+        if (typeof validationSeq !== "number" || !Number.isInteger(validationSeq) || validationSeq < 0) {
+          throw `Validation sequence number is not a positive integer: ${validationSeq}`;
+        }
+
+        if (typeof validationDigest !== "string" && !(validationDigest instanceof String)) {
+          throw "Validation digest is t a string.";
+        }
+
+        // Basic sanity checks passed, set the validation payload to be processed:
+        validationPayload = {
+          seq: validationSeq,
+          digest: validationDigest,
+        };
+
+        // TODO: typecheck!
+      }
+      consistencyLoopback = payload["consistencyLoopback"];
+    } catch (e) {
+      console.log("Error decoding received message:", decPayload, e);
+      return;
+    }
+
+    try {
+      // Now try to validate the message and potentially trim hash chains. This
+      // also implicitly handles loopback messages sent from our own device. It
+      // does not detect reordering for loopback messages:
+      let trimmed = this.messageChains.validate_trim_chain(
+        msg.sender,
+        (validationPayload) ? validationPayload.seq : null,
+        (validationPayload) ? validationPayload.digest : null,
+      );
+      console.log(`Trimmed ${trimmed} entries from pairwise message-chain with ${msg.sender}`);
+
+      // Finally, insert the message into the chain. This also validates
+      // ordering and contents for messages looped back to us.
+      this.messageChains.insert_message(msg.sender, messagePayload, recipients);
```

Comment:
need to sort onMessage as well - we can maybe let the server reorder recipients or at least detect when it does.